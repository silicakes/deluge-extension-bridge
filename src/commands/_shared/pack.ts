/**
 * Utilities for 7-bit packing and unpacking for MIDI SysEx.
 */

/**
 * Convert binary data to 7-bit MIDI-safe format (for SysEx).
 * @param data Uint8Array of data to encode.
 * @returns Encoded array of 7-bit values.
 */
export function encode7Bit(data: Uint8Array): number[] {
  const result: number[] = [];
  for (let i = 0; i < data.length; i += 7) {
    const bytesInGroup = Math.min(7, data.length - i);
    let highBits = 0;
    for (let j = 0; j < bytesInGroup; j++) {
      if (data[i + j] & 0x80) {
        highBits |= 1 << j;
      }
    }
    result.push(highBits);
    for (let j = 0; j < bytesInGroup; j++) {
      result.push(data[i + j] & 0x7f);
    }
  }
  return result;
}

/**
 * Decode 7-bit MIDI-safe format back to binary data.
 * @param encoded Array of 7-bit values with high-bits bytes.
 * @returns Original binary data as Uint8Array.
 */
export function decode7Bit(encoded: number[]): Uint8Array {
  const result: number[] = [];
  let i = 0;
  while (i < encoded.length) {
    const highBits = encoded[i++];
    const groupLen = Math.min(7, encoded.length - i);
    for (let j = 0; j < groupLen; j++) {
      const low7 = encoded[i++];
      const bit = (highBits >> j) & 0x01;
      result.push((bit << 7) | low7);
    }
  }
  return new Uint8Array(result);
}
