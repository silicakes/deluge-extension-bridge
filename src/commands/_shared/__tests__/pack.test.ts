import { encode7Bit, decode7Bit } from "../pack";
import { describe, it, expect } from "vitest";

describe("encode7Bit", () => {
  it("encodes data into 7-bit segments", () => {
    const data = new Uint8Array([
      0x00, 0x7f, 0x80, 0xff, 0x55, 0xaa, 0x01, 0x02,
    ]);
    const encoded = encode7Bit(data);
    expect(encoded).toMatchSnapshot();
  });
});

describe("decode7Bit", () => {
  it("decodes encoded data back to original", () => {
    const original = new Uint8Array([
      0x00, 0x7f, 0x80, 0xff, 0x55, 0xaa, 0x01, 0x02,
    ]);
    const encoded = encode7Bit(original);
    const decoded = decode7Bit(encoded);
    expect(decoded).toEqual(original);
  });
});
