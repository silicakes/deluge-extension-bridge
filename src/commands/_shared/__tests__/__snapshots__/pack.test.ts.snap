// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encode7Bit > encodes data into 7-bit segments 1`] = `
[
  44,
  0,
  127,
  0,
  127,
  85,
  42,
  1,
  0,
  2,
]
`;
